### ISSUE TITLE
GeoConnnect: Can map restricted files and restrict mapped files but anyone can download from WorldMap

### ISSUE THREAD

**Issue by kcondon** [IQSS/dataverse#3636]
While testing World Map, it was discovered that the generic username "dataverse_user" is displayed as a link on the map page. Clicking on it lists all mapped files generated by dataverse using that generic user. 

This was unexpected but not viewed as a problem because there is an assumption that all data is public -mapped data is required to be published and assumed to not be restricted. 

## Core Issue (Dataverse codebase change): 

"However you can map restricted files and restrict mapped files, giving the impression of controlled access. This is mostly true from the Dataverse side but not from the World Map side via the owner link. You can use the Download function to download the data in csv format."

## Fix

- Do not allow mapping of restricted files (even if formerly public). 
- Question: If a formerly unrestricted and published file has  a map, is that OK?  e.g. Should it be removed from the WorldMap?

**Comment by raprasad** [comment]
Answer to question at end of issue:

---
 
### **YES**, delete the newly restricted file from WorldMap

Upon a mapped file being restricted, Dataverse makes an API call to Geoconnect.
A new endpoint on Geoconnect will:

1.  Receive datafile id + installation name from Dataverse
1.  Make an API call to WorldMap to:
    -  Verify that this is a Dataverse-created layer on WorldMap (existing WorldMap endpoint)
1.  Deletes the layer (existing WorldMap endpoint)

Note: Step 1 above can be made into a two step process (slightly better way):

1. Dataverse sends a callback url to geoconnect with a token
1. Geoconnect uses the callback + token to retrieve the datafile id + installation name from Dataverse

**Comment by raprasad** [comment]
fyi: working on the geoconnect endpoint first

**Comment by raprasad** [comment]
This gist describes using a callback url for file deletion initiated by Dataverse and not using the Geoconnect UI:   https://gist.github.com/raprasad/210ed1dc852181db5033aa41c9e24b93

**Comment by pdurbin** [comment]
@sekmiller and I just talked through this issue. My thoughts:

- Like #3480 this should be an exceedingly **RARE EVENT**. We're talking about someone who has made data available for public download for a while and then changed their mind and decided to restrict that data. The horse is out of the barn at this point. What is the user experience for someone who is doing this? If we already have a popup explaining what they're doing, let's add "Oh, and we detected you created a map based on this previously public data, which we'll delete for you. The thumbnail is going away too." I question how much automation we really need here. @sekmiller has already hidden the "Map Data" button in 4d9bdc7 which is a great start.
- Since you can toggle restricted on and off all day long while the dataset is in draft, the action should take place in the `PublishDatasetCommand`.
- It makes sense to me to add a boolean such as `markedForDeletion` to the `maplayermetadata` table to record the intent to delete the map both from WorldMap and Dataverse (the row in that table and the thumbnails). It probably doesn't need to happen within seconds of republishing the dataset. Maybe Geoconnect could query Dataverse periodically for a list of maps that should be deleted from WorldMap? Maybe we could add the boolean to the `curl -X POST http://localhost:8080/api/admin/geoconnect/mapLayerMetadatas/check` API I added for #3480 and a sysadmin could delete the map from WorldMap via the WorldMap GUI and then run `curl -H "X-Dataverse-key: $API_TOKEN" http://localhost:8080/api/files/{file_id}/map` to delete the row from `maplayermetadata` and the thumbnails from the Dataverse side. I documented these API endpoints at https://github.com/IQSS/dataverse/blob/2ca06dca33e62a5a37c6ef1339e2a1d0b65c8229/doc/sphinx-guides/source/admin/geoconnect-worldmap.rst#removing-dead-explore-links

**Comment by pdurbin** [comment]
It looks like Dataverse doesn't currently explain anything consequences of restricting a previously public file. There's no warning that the horse is already out of the barn. Here's what the user experiences:

![screen shot 2017-04-11 at 3 47 55 pm](https://cloud.githubusercontent.com/assets/21006/24927917/8f828e02-1ece-11e7-9d89-06927b063bcd.png)

![screen shot 2017-04-11 at 3 48 11 pm](https://cloud.githubusercontent.com/assets/21006/24927925/93854436-1ece-11e7-9c4a-7ad03f5bb375.png)

![screen shot 2017-04-11 at 3 48 28 pm](https://cloud.githubusercontent.com/assets/21006/24927930/974a79c4-1ece-11e7-9523-378053f90b8a.png)

I assume this user story is not very common. For the matter at hand we might want to at least tell them that their map and thumnbnail will be deleted and why.

**Comment by pdurbin** [comment]
> It makes sense to me to add a boolean such as markedForDeletion to the maplayermetadata table to record the intent to delete the map both from WorldMap and Dataverse (the row in that table and the thumbnails).

On second thought, maybe this is redundant. From the row in the `maplayermetadata` table you can just get at the file and see if it's restricted or not. I still think it would be nice to add another `check` API endpoint to query the state of the Dataverse database to see if things are in a weird state, such as row in maplayermetadata existing when it's associated with a restricted file.

**Comment by raprasad** [comment]
It can be broken into:
1. **Message 1**: Upon file restriction, if a worldmap layer exists, warn user it will be deleted
1. **Message 2**: On the dataset publication confirmation step, list all restricted files with worldmap layers and warn the user that the map layers will be deleted
1. **Action**: On actual dataset publication, find all restricted files with worldmap layers delete them

### Important: Talk to UX team about (1) and (2)

**Comment by mheppler** [comment]
Added static delete WorldMap confirm popup to dataset pg and warning msg to File Restrict popups on both dataset and file pgs.

![screen shot 2017-04-13 at 9 44 31 am](https://cloud.githubusercontent.com/assets/687227/25008014/5eff1df0-2030-11e7-86e0-bc42173dbc81.png)
![screen shot 2017-04-13 at 9 59 20 am](https://cloud.githubusercontent.com/assets/687227/25008015/5f006458-2030-11e7-8444-f9883e60c3e0.png)

**Comment by dlmurphy** [comment]
I'd like to suggest these revisions to those popup messages:

Note: If you restrict access to this file, then once it's published, the file's associated map on WorldMap will be deleted and its "Explore on WorldMap" feature will no longer be accessible. 

Note: This dataset contains newly-restricted files with associated WorldMap maps. If you publish this dataset, these files will have their WorldMap maps deleted due to the file access changes in this dataset version. Do you want to continue?

**Comment by raprasad** [comment]
to do:
https://gist.github.com/raprasad/210ed1dc852181db5033aa41c9e24b93

**Comment by pdurbin** [comment]
As far as I understand it, below. is the main plan from the gist above to actually do the deletions from WorldMap and Dataverse. Since I added the `DeleteMapLayerMetadataCommand` I'm providing some more detail:

- [ ] (1) Dataverse connects to the Geoconnect "delete" API sending a (A) token and (B) callback url
  - New Geoconnect API endpoint--or existing endpoint with "delete" action
- [ ] (2) Geoconnect uses the token and a callback url to retrieve the Datafile id and Dataverse installation info from Dataverse
  - Update existing Dataverse API endpoint. (add "datafile_is_restricted" attribute)
     - DV retrieves datafile information related to token, makes sure it's restricted, and returns 'ok' with datafile id back to Geoconnect
- [ ] (3) Once Geoconnect receives the datafile id, Geoconnect will attempt to delete the file from WorldMap
     - Geoconnect attempts deletion with existing WorldMap endpoint
- [ ] (4) The results of the delete attempt will be sent back to to Dataverse using the token
   - Existing Dataverse endpoint: Attempt to delete the map layer metadata associated with the datafile. That endpoint is `curl -H "X-Dataverse-key: $API_TOKEN" -X DELETE http://localhost:8080/api/files/{file_id}/map` and requires an API token from a user who has `Permission.EditDataset` since what's being called is the `DeleteMapLayerMetadataCommand`. This endpoint is documented at https://github.com/IQSS/dataverse/blob/develop/doc/sphinx-guides/source/admin/geoconnect-worldmap.rst#removing-dead-explore-links

In addition, the static delete WorldMap confirm popups (screenshots above) are going to be made real:

- [ ] Wire up static delete WorldMap confirm popups

Questions:

- What if Geoconnect happens to be down when publishing from Dataverse?
- The token in step 4 is a regular Dataverse API token and has a good amount of power so can we remove it from Geoconnect as soon as it is no longer needed?

**Comment by landreev** [comment]
Regarding a question from @pdurbin 's last comment:

"What if Geoconnect happens to be down when publishing from Dataverse?"

- I still don't have a good answer for this... What should we do? I was considering making the change in the file restriction status *depend on the success* of deleting the file from WorldMap... i.e., if we failed to delete the file in WM, the user gets "sorry, could not restrict the file - try again later". But that's not going to work with the step 2) above - where Geoconnect retrieves the metadata for the file, and expects to confirm that it is listed as "restricted"... and that relies on the change to have already been made in the db. 

Any thoughts? A simple, low-cost alternative is a warning in the logs  and a notification sent to the admin; telling them that there's now a file in worldmap that somehow needs to be deleted.

**Comment by landreev** [comment]
Ok, that was a confused comment... I typed it as I was leaving - then realized I was mixing up 2 different updates, almost as soon as I was out the door... So, the map needs to be deleted when the dataset gets published, not when the file gets restricted...
So yeah, it should be possible to modify the PublishDatasetCommand to somehow wait for the confirmation, that the file was deleted on the WM side - and fail if it can't be deleted... But I'm not sure that's how we want to handle it... Let's chat about it quickly tomorrow...

**Comment by pdurbin** [comment]
@landreev it might be helpful for you to review the decision made by @sekmiller @raprasad and myself at https://iqss.slack.com/archives/G4D2TJ4K0/p1492715850824083 to go ahead and call the `DeleteMapLayerMetadataCommand` from the `PublishDatasetCommand`. This is a great simplification in terms of security since the person who has permission to publish is also able to delete the map layer. @sekmiller make this change in 56c1696.

**Comment by landreev** [comment]
@pdurbin - I know, I mentioned it to Raman earlier this week, that the step 4. in your outline above was unnecessary (since the whole thing was initiated on the dataverse side - so, easy to call the command on the dataverse side...) So, no, I wasn't going to make that API call from geoconnect. As you said, we would then have to worry about giving geoconnect a "real" API token, etc.

**Comment by landreev** [comment]
(p.s. note that the version of PublishDatasetCommand that @sekmiller checked in last week already had the DeleteMapLayerMetadataCommand in it)

**Comment by landreev** [comment]
The code is checked in. In two places - in Geoconnect (python) and on the Dataverse side. 
It needs to be reviewed; both the way I implement the work flow (it appears to be working, but I'm not sure it's the best way possible, the way I go about it). And also, where I stick the code, and how I name methods and access points. Especially on the GeoConnect side - I don't have a good feel yet for what's good style there... So let's review all this.  

Oh, and of course the issue I brought up 4 days ago. I picked the hardcore way of handling the condition - I bail out of PublishDataset if that layer data cannot be deleted. 
But it's probably *too* hardcore...

**Comment by raprasad** [comment]
Dataverse creating bad to call geoconnect:
```http://127.0.0.1:8070/tabular/delete-map-no-ui/http://127.0.0.1:8070/shapefile/map-it/f343469d35c8f47ff0a547f3c926a9f54d28737ea2aeedf5491abd90d80fcfe4/?cb=https%3A%2F%2F127.0.0.1%2F%2Fapi%2Fworldmap%2Fdatafile%2F```
Info:   response code: 404

**Comment by raprasad** [comment]
- no extra warning messages seem to be appearing when restricting a tabular file

**Comment by raprasad** [comment]
dataverse code: callback url does not have port

## fails locally, no port

```
        String callback_url =  URLEncoder.encode(systemConfig.getDataverseSiteUrl() + "/" + WorldMapRelatedData.GET_WORLDMAP_DATAFILE_API_PATH);
```

## example of getting url with port

```
        String callback_url = this.getServerNamePort(request) + GET_WORLDMAP_DATAFILE_API_PATH;
```
```
this.getServerNamePort(request) 
```

**Comment by raprasad** [comment]
### bug: 
- the publish command* does not make any database commits until the command completes
- the delete worldmap layer command requires the creation of a callback_url that includes a _new_ token that has been _commited to the db_
- since the token isn't saved to the db, the geoconnect callback fails to find the token and the delete fails

(*by design) 

### workaround

- discussed potential workaround with @sekmiller of pre-creating any tokens before the publish command is called ... checking viability of this approach

---

db transaction that is being "held back" until end of the command.
  - See below ```return this.save(token);``` 
  - The ```WorldMapToken``` object doesn't receive an ```id``` until after the command completes -- which is too late for the callback

```java
/**
     * 
     * Retrieve a fresh token to map a Dataverse file on GeoConnect
     * 
     * @param dataFileID
     * @param dvUserID
     * @return WorldMapToken
     */
    public WorldMapToken getNewToken(DataFile dataFile, AuthenticatedUser dvUser){
        if ((dataFile==null)||(dvUser==null)){
            logger.severe("dataFile or dvUser is null");
            return null;
        }
        WorldMapToken token = new WorldMapToken();
        token.setApplication(tokenApplicationService.getGeoConnectApplication());
        token.setDatafile(dataFile);
        token.setDataverseUser(dvUser);
        token.setModified();
        token.setToken();   
        return this.save(token);
    }
```

**Comment by raprasad** [comment]
@landreev : thanks for writing.  I'll hold back and think about this.  The only real reason for the token is to stop arbitrary delete calls to _geoconnect_.  
  - e.g. the initial call to geoconnect doesn't even have the datafile id in the url, only the token connected to the datafile id.

>  But, wouldn't it be much easier to modify the Dataverse API (the one that the callback URL calls), to return the restricted status in the most recent version - and not in the published version? Then it would just work...

Right now that API endpoint does start with the DataFile Id and then calls ```getLatestFileMetadata``` so I think it will return the state of the draft

> But, you know, it sounds like work.

Yes, a reason to sleep on it--especially since it adds extra complexity _before_ calling the command--the reasons for which will likely be lost over time, add tech. debt, etc.

---

### the way to go?

Back to your initial paragraph.  Yes, if delete was done upon restricting the file, then it would work as is -- and the user warning *is* already there. 

  - **support** for delete on restrict: Creating a map, which includes adding metadata to the dataset, does *not* bump the dataset version level.  
    - e.g. so deletion w/o publishing would be consistent

**Comment by landreev** [comment]
OK, if doing this when the file gets restricted doesn't add extra complexity - then why not... I mean, yes, we'll just need to make that decision. 
But getting the token saved in the database right away should by itself be easily achievable, by adding @TransactionAttribute(REQUIRES_NEW) to the getNewToken() method in the token service bean. 

(btw, is WorldMap is up, this delete operation should never take long - is this a correct assumption?)

**Comment by landreev** [comment]
(sorry if any of it was way off - i was trying to understand what was going on while also trying to pay attention to somebody talking about Keystone)

**Comment by raprasad** [comment]
geoconnect pull request to review along with DV code:  https://github.com/IQSS/geoconnect/pull/132

**Comment by landreev** [comment]
Kevin reported a problem with the web error notification, for a failed delete of a layer. (the email notification is working as it should). 
I'm looking into it.

**Comment by kcondon** [comment]
After restricting a previously mapped shape file, going to the file landing page has a broken image and throws a null ptr in logs

**Comment by kcondon** [comment]
OK, the basic functionality is working for all map types. Leonid fixed the notifications and now we are waiting on the null ptr file landing page fix.

**Comment by raprasad** [comment]
@kcondon fix for null pointer is in.  may be QA'ed

**Comment by kcondon** [comment]
![screen shot 2017-05-19 at 11 49 44 am](https://cloud.githubusercontent.com/assets/4983816/26256087/c76a6bfe-3c89-11e7-9983-fdac83198f79.png)

Removing map layer using geoconnect remove button when returning to dataverse fails with above error, though actual map layer is deleted.
